### PR TITLE
Improve arrow function parsing

### LINKED ISSUE #255: Can't call conditional expression
```ts
function f(x: i32): i32 { return x; }
function g(x: i32): i32 { return x + 2; }
export function test(b: boolean): i32 {
  return (b ? f : g)(1);
}
```

Currently fails with: 
```
ERROR TS1005: ')' expected.

   return (b ? f : g)(1);
             ~
```

### PR DESCRIPTION
**Pull request by andy-hanson**
Fixes parsing for #255. Does not fix the issue since a conditional expression is still unsupported in the resolver.

Previously at `(a ? b : c)()`, we would assume that `a ? b : c` were the parameters of an arrow function because every token involved was a valid token to appear in a signature -- but that wasn't strict enough. Now we will verify that `?` is always followed by `:`, `,`, or `)`, none of which are legal in expressions.
Also correctly parses `x => x` now, though it still won't compile.

### CODE REVIEW

File: src/parser.ts [hunk 1]
```diff
@@ -1420,6 +1420,11 @@ export class Parser extends DiagnosticEmitter {
     return Node.createFunctionExpression(declaration);
   }
 
+  private parseArrowFunctionBody(tn: Tokenizer): Statement | null {
+    // TODO: should parse an expression unless the next token is `{`.
```
Comment (dcodeIO) [inline src/parser.ts]: `tn.peek()` can check the next token, would that help?

File: src/compiler.ts [hunk 2]
```diff
@@ -5445,7 +5444,9 @@ export class Compiler extends DiagnosticEmitter {
         }
       }
     } else {
-      body.push(this.compileStatement(bodyStatement));
+      body.push(bodyStatement instanceof Expression
```
Comment (dcodeIO) [inline src/compiler.ts]: This should probably be `bodyStatement.kind != NodeKind.BLOCK`, as `instanceof` isn't supported (yet, without a GC).

**Comment by dcodeIO** [comment]
The commit above does something very similar with a somewhat reduced refactoring footprint. Also fixes related serializer issues but doesn't tackle `x => x`. We should probably support this as well eventually, even if just for parsing and erroring out (no type information the compiler could utilize).

**Comment by andy-hanson** [comment]
Closing in favor of referenced commit.

**Pull request closed without merging**